### ISSUE TITLE
Support Java 9 Multi-Release JAR Files

### ISSUE THREAD

**Issue by marchof** [jacoco/jacoco#407]
Java 9 has support for multiple class file versions in the same JAR file. As of today this will lead to an exception in JaCoCo. We need to find a way how JaCoCo handles such JAR files. See:

http://openjdk.java.net/jeps/238

As a starting point I would simply filter all classes from META-INF/versions folders within JARs.

**Comment by Godin** [comment]
@marchof just to confirm - I suppose you want to implement this filtering, because it can't be done in configuration by user?

**Comment by marchof** [comment]
@Godin Right, we can't filter from inside a Jar (this might be another story).

I don't know how soon we will see multi-release Jars. That's more like a reminder.

**Comment by Godin** [comment]
Maybe we'll see multi-release JARs soon - https://issues.apache.org/jira/browse/LUCENE-7966

**Comment by igr** [comment]
See the above link for the workaround

**Comment by lagergren** [comment]
So is this ever going to get any attention? I'm trying to work around this with excludes, but it seems JaCoCo reads and parses files and report this error before checking any of the exclude filters (at least I think that's why - and that's also rather counterintuitive)

Thanks!

**Comment by Godin** [comment]
@lagergren exclusion of `META-INF/**` works just fine - https://stackoverflow.com/a/50294239/244993

**Comment by lagergren** [comment]
Actually, Goden, it doesn't, because of various inherit complexity in our build process. We need to do various work with the compiled classes before they are JARed. A simple way to fix this would be to make JaCoCo aware of excludes and includes before actually scanning the file tree (and then bailing on classes with the same names). Here it would just consider that file excluded.

The current case is different compiled versions of the same class, e.g. HelloWorld.class, one for 8 and one for 11 in separate directories. It's already too late for JaCoCo then.

**Comment by Godin** [comment]
@lagergren sorry, but I don't understand what you are trying to say, because exclusions already happen before analysis - https://github.com/jacoco/jacoco/blob/v0.8.2/jacoco-maven-plugin/src/org/jacoco/maven/ReportSupport.java#L198-L201

**Comment by antoinetran** [comment]
Hm the exclude workaround didn't work for us, but the error was not at jacoco level but at maven-project-info-reports-plugin level. We use 2.9, but the latest 3.0.0 has a blocking regression for us: https://issues.apache.org/jira/browse/MPIR-373.
So we stayed at 2.9 but upgraded Bcel, because this is the one reporting the error. It was fixed in https://issues.apache.org/jira/browse/BCEL-300

Workaround:
```
<plugin>
	<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-project-info-reports-plugin</artifactId>
	<version>2.9</version>
	<dependencies>
		<!-- See https://github.com/jacoco/jacoco/issues/407
		and https://issues.apache.org/jira/browse/BCEL-300
		This is because of Java9 byte-code instrumentation (even for Java8 compiler). -->
		<dependency>
			<groupId>org.apache.bcel</groupId>
			<artifactId>bcel</artifactId>
			<version>6.4.1</version>
		</dependency>
	</dependencies>
</plugin>
```

**Comment by djchapm** [comment]
Would like to bump this issue - now with JDK-11 in sunset and JDK-17 the official LTS - people will probably run into this again here in 2022.  Creating multi-release jars is the most convenient solution to providing libraries and being able to support jacoco code coverage is still a great benefit.

**Comment by marchof** [comment]
@djchapm Why does the `META-INF/**` exclusion not work for you?

**Comment by Godin** [comment]
@kwin `excludes` parameter of `report` mojo is about files in the `target/classes` directory (https://github.com/jacoco/jacoco/blob/v0.8.2/jacoco-maven-plugin/src/org/jacoco/maven/ReportSupport.java#L198-L201), so in your case (`target/classes/jaxb-api-2.3.1.jar@META-INF/versions/9/javax/xml/bind/ModuleUtil.class`) you can try `**/*.jar` which should exclude all jar files.
